### PR TITLE
feature: Local noise simulator

### PR DESCRIPTION
**Pull request by yitinche**
Description of changes:
- Add density matrix simulation method for noise simulation.
- Add noise operations and density matrix result type.
- Break down DefaultSimulator() into DefaultSimulator() and StateVectorSimulator() . The former is inherited by StateVectorSimulator and DensityMatrixSimulator.
- Break down Simulation() into Simulation() and StateVectorSimulation() . The former is inherited by StateVectorSimulation and DensityMatrixSimulation.

Build directory: [build_files.tar.gz](https://github.com/aws/amazon-braket-default-simulator-python/files/5104544/build_files.tar.gz)

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ x ] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [ x ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ x ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ x ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ x ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yitinche** [comment]
> ### AWS CodeBuild CI Report
> * CodeBuild project: DefaultSimulatorPullRequestBuild
> * Commit ID: [5271899](https://github.com/aws/amazon-braket-default-simulator-python/commit/5271899a75a39866b230d9120a70d9930bd9144c)
> * Result: FAILED
> * [Build Logs](https://9k923l77sk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=946c3af3-f6bd-46c9-8177-a7ce52b72524%2Fbuild.log) (available for 30 days)
> 
> _Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)_

The CodeBuild CI fails because this repo has dependency on the "noise" branch of Schemas repo. I have checked that the CI will pass with the noise branch of Schemas.

File: src/braket/default_simulator/noise_operations.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        return self._targets
+
+
+@_from_braket_instruction.register(braket_instruction.Amplitude_Damping)
+def _amplitude_damping(instruction) -> Amplitude_Damping:
+    return Amplitude_Damping([instruction.target], instruction.prob)
+
+
+class Kraus(KrausOperation):
+    """Arbitrary quantum channel"""
```
Comment (licedric) [inline src/braket/default_simulator/noise_operations.py:112]: Like in the SDK, it would be great if you could describe the actual channel that's defined by the matrices.

File: src/braket/default_simulator/operation_helpers.py [hunk 2]
```diff
@@ -92,6 +114,20 @@ def check_hermitian(matrix: np.ndarray):
...
 
 
+def check_CPTP(matrices: List[np.ndarray]):
+    """ Checks that the given matrices define a CPTP map.
+
+    Args:
+        matrices (List[np.ndarray]): The matrices to check
+
+    Raises:
+        ValueError: If the matrices do not define CPTP a CPTP map
```
Comment (licedric) [inline src/braket/default_simulator/operation_helpers.py:124]: Typo

File: src/braket/default_simulator/operation_helpers.py [hunk 3]
```diff
@@ -92,6 +114,20 @@ def check_hermitian(matrix: np.ndarray):
...
+    """
+    E = sum([np.dot(matrix.T.conjugate(), matrix) for matrix in matrices])
+    if not np.allclose(E, np.eye(*E.shape)):
+        raise ValueError(f"{matrices} does not satisfy CPTP")
```
Comment (licedric) [inline src/braket/default_simulator/operation_helpers.py:128]: does not satisfy CPTP -> do not define a CPTP map

File: src/braket/default_simulator/operation_helpers.py [hunk 4]
```diff
@@ -92,6 +114,20 @@ def check_hermitian(matrix: np.ndarray):
...
+def check_CPTP(matrices: List[np.ndarray]):
+    """ Checks that the given matrices define a CPTP map.
+
+    Args:
+        matrices (List[np.ndarray]): The matrices to check
+
+    Raises:
+        ValueError: If the matrices do not define CPTP a CPTP map
+    """
+    E = sum([np.dot(matrix.T.conjugate(), matrix) for matrix in matrices])
```
Comment (licedric) [inline src/braket/default_simulator/operation_helpers.py:126]: Use np.matmul instead of np.dot for clarity.

File: src/braket/default_simulator/simulation.py [hunk 5]
```diff
@@ -76,71 +47,7 @@ def evolve(self, operations: List[GateOperation]) -> None:
...
-        """
-        np.ndarray: The final state vector of the simulation after application of observables.
-
-        Raises:
-            RuntimeError: If observables have not been applied
-        """
-        if self._post_observables is None:
-            raise RuntimeError("No observables applied")
-        return self._post_observables
+        raise NotImplementedError("evolve has not been implemented yet.")
```
Comment (licedric) [inline src/braket/default_simulator/simulation.py:50]: I'd use present tense: "evolve is not implemented". Ditto below.

File: src/braket/default_simulator/simulator.py [hunk 6]
```diff
@@ -43,7 +41,7 @@ class DefaultSimulator(BraketSimulator):
     DEVICE_ID = "default"
 
     def run(
-        self, circuit_ir: Program, qubit_count: int, shots: int = 0, *, batch_size: int = 1,
+        self, circuit_ir: Program, qubit_count: int, shots: int, simulation: Simulation,
```
Comment (licedric) [inline src/braket/default_simulator/simulator.py:44]: I would be extremely careful about making a change like this, since it looks like a breaking change. You need to be extremely careful about making sure that existing code doesn't break.

In particular, if you make this change, does the code in the SDK still work? I.e. when I create a LocalSimulator in the SDK, does it have the same behavior as the old code?

File: src/braket/default_simulator/simulator.py [hunk 7]
```diff
@@ -66,14 +60,11 @@ def run(
                 as a result type when shots=0. Or, if statevector and amplitude result types
                 are requested when shots>0.
 
-
         Examples:
             >>> circuit_ir = Circuit().h(0).to_ir()
             >>> DefaultSimulator().run(circuit_ir, qubit_count=1, shots=100)
```
Comment (licedric) [inline src/braket/default_simulator/simulator.py:65]: Does this example actually work without specifying Simulation?

File: src/braket/default_simulator/simulator.py (hunk 6 above)
Comment (licedric) [reply src/braket/default_simulator/simulator.py:44]: A better way of doing this ideally would be to keep the existing implementation of DefaultSimulator, and add another DensityMatrixSimulator, with different backend type.

File: src/braket/default_simulator/statevector_simulation.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
+# Copyright 2019-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (licedric) [inline src/braket/default_simulator/statevector_simulation.py:1]: Name it state_vector_simulation.py instead.

File: src/braket/default_simulator/densitymatrix_simulation.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
+# Copyright 2019-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (licedric) [inline src/braket/default_simulator/densitymatrix_simulation.py:1]: Name it density_matrix_simulation.py instead.

File: test/unit_tests/braket/default_simulator/test_statevector_simulation.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
+# Copyright 2019-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (licedric) [inline test/unit_tests/braket/default_simulator/test_statevector_simulation.py:1]: Can you also test the behavior of attempting to simulate a circuit with noise?

File: src/braket/default_simulator/densitymatrix_simulation_helpers.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
+# Copyright 2019-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (kshitijc) [inline src/braket/default_simulator/densitymatrix_simulation_helpers.py:1]: I'm not sure why we need a separate *helpers module. We should try and reduce the number of such "helpers" files. They tend to transform into a large collection of unrelated code over time, and don't end up being any help. I realize that we already have quite a few of these helpers in our packages, but we should try to avoid creating new ones if possible.

File: setup.py [hunk 12]
```diff
@@ -27,7 +27,12 @@
     packages=find_namespace_packages(where="src", exclude=("test",)),
     package_dir={"": "src"},
     install_requires=["amazon-braket-schemas", "numpy", "opt_einsum"],
-    entry_points={"braket.simulators": ["default = braket.default_simulator:DefaultSimulator"]},
+    entry_points={
+        "braket.simulators": [
+            "default = braket.default_simulator.statevector_simulator:StateVectorSimulator",
+            "noise_simulator = braket.default_simulator:DensityMatrixSimulator",
```
Comment (kshitijc) [inline setup.py:33]: I'm not sure why we are using inconsistent object references for the two.

File: test/unit_tests/braket/default_simulator/test_statevector_simulation.py (hunk 10 above)
Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_statevector_simulation.py:1]: when when using state vector simulator with noise, it throws TypeError. Example: `TypeError: instruction BitFlip is not supported by DefaultSimulator`

File: src/braket/default_simulator/density_matrix_simulation.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    ) -> np.ndarray:
+        """ Apply a matrix M to a density matrix D according to:
+
+            .. math::
+                D \rightarrow M D M^{\\dagger}
+
+        Args:
+            state (np.ndarray): initial density matrix
+            qubit_count (int): number of qubit in the circuit
+            matrix (nd.array): matrix to be applied to the density matrix
```
Comment (licedric) [inline src/braket/default_simulator/density_matrix_simulation.py:182]: np.ndarray

File: src/braket/default_simulator/density_matrix_simulation.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+            targets (Tuple[int,...]): qubits of the density matrix the matrix applied to.
+
+        Returns:
+            state (nd.ndarray): output density matrix
```
Comment (licedric) [inline src/braket/default_simulator/density_matrix_simulation.py:186]: np.ndarray

File: src/braket/default_simulator/noise_operations.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    """Phase Damping noise channel"""
+
+    def __init__(self, targets, probability):
+        self._targets = tuple(targets)
+        self._probability = probability
+
+    @property
+    def matrices(self) -> np.ndarray:
+        K0 = np.array([[1.0, 0.0], [0.0, np.sqrt(1 - self._probability)]], dtype=complex)
+        K1 = np.array([[0.0, np.sqrt(self._probability)], [0.0, 0.0]], dtype=complex)
```
Comment (licedric) [inline src/braket/default_simulator/noise_operations.py:123]: This defines amplitude damping, not phase damping.

File: src/braket/default_simulator/state_vector_simulator.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        Args:
+            circuit_ir (Program): ir representation of a braket circuit specifying the
+                instructions to execute.
+            qubit_count (int): The number of qubits to simulate.
+            shots (int): The number of times to run the circuit.
+            batch_size (int): The size of the circuit partitions to contract,
+                if applying multiple gates at a time is desired; see `StateVectorSimulation`.
+                Must be a positive integer.
+                Defaults to 1, which means gates are applied one at a time without any
+                optmized contraction.
```
Comment (licedric) [inline src/braket/default_simulator/state_vector_simulator.py:41]: optimized

File: src/braket/default_simulator/state_vector_simulator.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class DefaultSimulator(BaseLocalSimulator):
+    def run(
+        self, circuit_ir: Program, qubit_count: int, shots: int = 0, batch_size: int = 1,
+    ) -> GateModelTaskResult:
+        """ Executes the circuit specified by the supplied `circuit_ir` on the simulator.
+
+        Args:
+            circuit_ir (Program): ir representation of a braket circuit specifying the
```
Comment (licedric) [inline src/braket/default_simulator/state_vector_simulator.py:33]: IR; Braket

File: test/unit_tests/braket/default_simulator/test_noise_operations.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+# language governing permissions and limitations under the License.
+
+import pytest
+
+import braket.ir.jaqcd as instruction
+from braket.default_simulator import noise_operations
+from braket.default_simulator.operation_helpers import check_cptp, from_braket_instruction
+from braket.ir.jaqcd import shared_models
+
+testdata = [
```
Comment (licedric) [inline test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: Can you add some test cases with probability = 0? with probability = 1? with probability = float("nan")? What happens with invalid Kraus matrices (e.g. 3x3)? What happens when targets size doesn't match the matrix size for Kraus matrices?

File: src/braket/default_simulator/densitymatrix_simulator.py [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+        simulation = DensityMatrixSimulation(qubit_count, shots)
+
+        return DefaultSimulator.run(self, circuit_ir, qubit_count, shots, simulation)
+
+    @property
+    def properties(self) -> GateModelSimulatorDeviceCapabilities:
+        observables = ["X", "Y", "Z", "H", "I", "Hermitian"]
+        max_shots = sys.maxsize
+        qubit_count = 26
```
[the remark this answers is not in the document]
Comment (yitinche) [reply src/braket/default_simulator/densitymatrix_simulator.py]: The 12 qubits took 10 seconds.
The 13 qubits took 70 seconds.
The 14 qubits took 690 seconds. 
So I think I will set the qubit_count to 13.

File: test/unit_tests/braket/default_simulator/test_noise_operations.py (hunk 18 above)
Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: Do we expect probability=1 work? Now it fails for probability=1 but works for probability=1.0

Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: Also, @avawang1 @kshitijc , do we want to add checks in local simulator repo? I saw GateOperation have no checks for angle and target, so I didn't add checks for KrausOperation neither. All checks are made in SDK and IR

Comment (avawang1) [reply test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: There may be a (not likely) scenario where the user does not use the SDK and just passes in a not-validated ir using the `construct` function in Pydantic. If you have time, feel free to add checks in this repo as well.

Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: @avawang1 To add checks to angle and probability, we might need to add classes for AngleGate and ProbabilisticNoise like we do in SDK. But I think it's would be good that we don't add more codes to the pull request because it's already big. So can we add checks later after this pull request?

Comment (avawang1) [reply test/unit_tests/braket/default_simulator/test_noise_operations.py:21]: Sure, np

File: src/braket/default_simulator/density_matrix_simulation.py [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+                if isinstance(operation, (GateOperation, Observable)):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_gate(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
+                if isinstance(operation, KrausOperation):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_kraus(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
```
Comment (speller26) [inline src/braket/default_simulator/density_matrix_simulation.py:116]: Is this supported for Kraus operations?

File: src/braket/default_simulator/density_matrix_simulation.py [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+    def probabilities(self) -> np.ndarray:
+        """np.ndarray: The probabilities of each computational basis state."""
+        return self.probabilities_from_state(self._density_matrix)
+
+    @staticmethod
+    def probabilities_from_state(state) -> np.ndarray:
+        """np.ndarray: The probabilities of each computational basis state of a
+        given state.
+        """
+        return np.real(np.diag(state))
```
Comment (speller26) [inline src/braket/default_simulator/density_matrix_simulation.py:169]: You can just put this into `probabilities`:

```python
@property
def probabilities(self) -> np.ndarray:
    """..."""
    return np.real(np.diag(self._density_matrix))
```

**Review by speller26 — COMMENTED** [review]
Remove the .DS_STORE file

File: src/braket/default_simulator/density_matrix_simulation.py [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+        np.ndarray: The current state of the simulation.
+        """
+        return self.density_matrix
+
+    @property
+    def density_matrix(self) -> np.ndarray:
+        """
+        np.ndarray: The density matrix specifying the current state of the simulation.
+        """
+        return self._density_matrix
```
Comment (kshitijc) [inline src/braket/default_simulator/density_matrix_simulation.py:145]: Some users might find this confusing (I did). Could you please elaborate in the docstrings on how they would be different?

File: src/braket/default_simulator/state_vector_simulator.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+    GateModelSimulatorDeviceCapabilities,
+    GateModelSimulatorDeviceParameters,
+)
+from braket.ir.jaqcd import Program
+from braket.task_result import GateModelTaskResult
+
+
+class DefaultSimulator(BaseLocalSimulator):
+    def run(
+        self, circuit_ir: Program, qubit_count: int, shots: int = 0, batch_size: int = 1,
```
Comment (kshitijc) [inline src/braket/default_simulator/state_vector_simulator.py:28]: This appears to change the existing behavior. `batch_size` could only be passed in as a kwarg earlier I think.

File: src/braket/default_simulator/simulation.py [hunk 24]
```diff
@@ -160,4 +67,11 @@ def shots(self) -> int:
     @property
     def probabilities(self) -> np.ndarray:
         """np.ndarray: The probabilities of each computational basis state."""
-        return np.abs(self._state_vector) ** 2
+        raise NotImplementedError("probabilities is not implemented.")
+
+    @staticmethod
+    def probabilities_from_state(state) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (yitinche) [reply src/braket/default_simulator/simulation.py:73]: The case where probabilities() doesn't work is when we want to get the probabilities from self._post_observables. The current implementation is hard-coded "np.abs(state) ** 2" (see the [code](https://github.com/aws/amazon-braket-default-simulator-python/blob/4e36b1b39ac155f71be2a988886de997d8596fe9/src/braket/default_simulator/result_types.py#L347) here), which is not compatible for density matrix. So, the probabilities_from_state() is for this purpose. 

On the other hand, the probabilities() is a more intuitive way to access probabilities() from native quantum state (ie. without observables).

File: test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py [hunk 25]
```diff
@@ -0,0 +1,426 @@
...
+
+from braket.default_simulator.density_matrix_simulator import DensityMatrixSimulator
+from braket.device_schema.simulators import (
+    GateModelSimulatorDeviceCapabilities,
+    GateModelSimulatorDeviceParameters,
+)
+from braket.ir.jaqcd import Program
+from braket.task_result import AdditionalMetadata, ResultTypeValue, TaskMetadata
+
+
```
[the remark this answers is not in the document]
Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py:29]: grcs_16 takes too long for density matrix, I created a grcs_8 test

File: src/braket/default_simulator/simulation.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (yitinche) [reply src/braket/default_simulator/simulation.py]: Removed probabilities_from_state() from simulation.py

File: test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py [hunk 26]
```diff
@@ -0,0 +1,426 @@
+# Copyright 2019-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py:1]: This is a good point. DensityMatrixSimulator is not supposed to use StateVectorSimulation. Do you have any idea to prevent this? 

For now, the easiest way seems to be to add a check that only allows DensityMatrixSimulator to use DensityMatrixSimulation, and StateVectorSimulator to use StateVectorSimulation. When there are more simulation methods for each simulator, we can add attributes and method to allow simulators to use different simulation methods.

File: src/braket/default_simulator/density_matrix_simulation.py (hunk 21 above)
Comment (yitinche) [reply src/braket/default_simulator/density_matrix_simulation.py]: We still need to expose a method to get probabilities from _post_observables.

I removed `probabilities_from_state` from simulation.py, state_vector_simulation.py, density_matrix_simulation.py 
and added this method as `_probabilities` in state_vector_simulation.py and density_matrix_simulation.py

File: test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (kshitijc) [reply test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py:1]: Adding tests here specific to the type of simulation that the simulator performs might help. For example, a test for noise simulation. Ideally, you'd want a set of tests which capture the complete functionality you expect from the class, and not tie it to the internal implementation (if possible).

[the remark this answers is not in the document]
Comment (yitinche) [reply test/unit_tests/braket/default_simulator/test_density_matrix_simulator.py:1]: Thanks, but my point is, in a fundamental level, density matrix simulator don't use state vector simulation. But the reverse is valid, state vector simulator can use density matrix simulation because it's more general.

**Comment by speller26** [comment]
Completed in #63

**Comment by speller26** [comment]
Completed in #63
